ICE: not enough bound vars

<!--
Thank you for finding an Internal Compiler Error! 🧊  If possible, try to provide
a minimal verifiable example. You can read "Rust Bug Minimization Patterns" for
how to create smaller examples.
http://blog.pnkfx.org/blog/2019/11/18/rust-bug-minimization-patterns/
-->

### Code

```Rust
async fn str<T>(T: &str) -> &str { &str }

```


### Meta
<!--
If you're using the stable version of the compiler, you should also check if the
bug also exists in the beta or nightly versions.
-->

`rustc --version --verbose`:
```
8cca42a47f5d574c8f7302c98c3f918cdc772fbb
```

### Error output

```
error[E0670]: `async fn` is not permitted in Rust 2015
 --> bad.rs:1:1
  |
1 | async fn str<T>(T: &str) -> &str { &str }
  | ^^^^^ to use `async fn`, switch to Rust 2018 or later
  |
  = help: pass `--edition 2021` to `rustc`
  = note: for more on editions, read https://doc.rust-lang.org/edition-guide

error[E0601]: `main` function not found in crate `bad`
 --> bad.rs:1:42
  |
1 | async fn str<T>(T: &str) -> &str { &str }
  |                                          ^ consider adding a `main` function to `bad.rs`
```

<!--
Include a backtrace in the code block by setting `RUST_BACKTRACE=1` in your
environment. E.g. `RUST_BACKTRACE=1 cargo build`.
-->
<details><summary><strong>Backtrace</strong></summary>
<p>

```

error: internal compiler error: compiler/rustc_middle/src/ty/visit.rs:434:21: Not enough bound vars: BoundRegion { var: 0, kind: BrNamed(DefId(0:6 ~ bad[f6ab]::str::'_), '_) } not found in []

thread 'rustc' panicked at 'Box<dyn Any>', /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_errors/src/lib.rs:1644:9
stack backtrace:
   0:     0x7f4813d74a01 - std::backtrace_rs::backtrace::libunwind::trace::h57c700f2e7064cc9
                               at /home/matthias/vcs/github/rust_debug_assertions/library/std/src/../../backtrace/src/backtrace/libunwind.rs:93:5
   1:     0x7f4813d74a01 - std::backtrace_rs::backtrace::trace_unsynchronized::h3bd6f8115b3aef84
                               at /home/matthias/vcs/github/rust_debug_assertions/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x7f4813d74a01 - std::sys_common::backtrace::_print_fmt::h4e3cd4a1386ea18b
                               at /home/matthias/vcs/github/rust_debug_assertions/library/std/src/sys_common/backtrace.rs:65:5
   3:     0x7f4813d74a01 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::ha4399f941a901f68
                               at /home/matthias/vcs/github/rust_debug_assertions/library/std/src/sys_common/backtrace.rs:44:22
   4:     0x7f4813debe48 - core::fmt::write::h3693e9c93ce31d95
                               at /home/matthias/vcs/github/rust_debug_assertions/library/core/src/fmt/mod.rs:1232:17
   5:     0x7f4813d3dc3f - std::io::Write::write_fmt::h7cd08f6496a9e03d
                               at /home/matthias/vcs/github/rust_debug_assertions/library/std/src/io/mod.rs:1684:15
   6:     0x7f4813d747f5 - std::sys_common::backtrace::_print::h1ea264f371a6489b
                               at /home/matthias/vcs/github/rust_debug_assertions/library/std/src/sys_common/backtrace.rs:47:5
   7:     0x7f4813d747f5 - std::sys_common::backtrace::print::h43e663fef4567fcf
                               at /home/matthias/vcs/github/rust_debug_assertions/library/std/src/sys_common/backtrace.rs:34:9
   8:     0x7f4813d3666e - std::panicking::default_hook::{{closure}}::h94496c5e3b8314dc
   9:     0x7f4813d3666e - std::panicking::default_hook::h62ce584520972394
                               at /home/matthias/vcs/github/rust_debug_assertions/library/std/src/panicking.rs:286:9
  10:     0x7f4816a24d42 - <alloc[88ef48cbf1494855]::boxed::Box<dyn for<'a, 'b> core[3a7799eb5b53fe42]::ops::function::Fn<(&'a core[3a7799eb5b53fe42]::panic::panic_info::PanicInfo<'b>,), Output = ()> + core[3a7799eb5b53fe42]::marker::Sync + core[3a7799eb5b53fe42]::marker::Send> as core[3a7799eb5b53fe42]::ops::function::Fn<(&core[3a7799eb5b53fe42]::panic::panic_info::PanicInfo,)>>::call
                               at /home/matthias/vcs/github/rust_debug_assertions/library/alloc/src/boxed.rs:2002:9
  11:     0x7f4816a24d42 - rustc_driver_impl[c3d665e357a167ae]::DEFAULT_HOOK::{closure#0}::{closure#0}
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_driver_impl/src/lib.rs:1160:17
  12:     0x7f4813d3715b - <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call::h20060c80854d9f26
                               at /home/matthias/vcs/github/rust_debug_assertions/library/alloc/src/boxed.rs:2002:9
  13:     0x7f4813d3715b - std::panicking::rust_panic_with_hook::hf3310e5853e10814
                               at /home/matthias/vcs/github/rust_debug_assertions/library/std/src/panicking.rs:692:13
  14:     0x7f4817c9ec23 - std[a3e1a1d40e09fad4]::panicking::begin_panic::<rustc_errors[7deb60bbfdd1de15]::ExplicitBug>::{closure#0}
                               at /home/matthias/vcs/github/rust_debug_assertions/library/std/src/panicking.rs:608:9
  15:     0x7f4817c999f6 - std[a3e1a1d40e09fad4]::sys_common::backtrace::__rust_end_short_backtrace::<std[a3e1a1d40e09fad4]::panicking::begin_panic<rustc_errors[7deb60bbfdd1de15]::ExplicitBug>::{closure#0}, !>
                               at /home/matthias/vcs/github/rust_debug_assertions/library/std/src/sys_common/backtrace.rs:137:18
  16:     0x7f4817bae6d6 - std[a3e1a1d40e09fad4]::panicking::begin_panic::<rustc_errors[7deb60bbfdd1de15]::ExplicitBug>
                               at /home/matthias/vcs/github/rust_debug_assertions/library/std/src/panicking.rs:607:12
  17:     0x7f4817c52356 - std[a3e1a1d40e09fad4]::panic::panic_any::<rustc_errors[7deb60bbfdd1de15]::ExplicitBug>
                               at /home/matthias/vcs/github/rust_debug_assertions/library/std/src/panic.rs:61:5
  18:     0x7f4817c4d6ca - <rustc_errors[7deb60bbfdd1de15]::HandlerInner>::bug::<&alloc[88ef48cbf1494855]::string::String>
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_errors/src/lib.rs:1644:9
  19:     0x7f4817c4d170 - <rustc_errors[7deb60bbfdd1de15]::Handler>::bug::<&alloc[88ef48cbf1494855]::string::String>
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_errors/src/lib.rs:1059:9
  20:     0x7f4817d4c037 - rustc_middle[ebb8a7eed001d673]::util::bug::opt_span_bug_fmt::<rustc_span[bb644a7833b5c27d]::span_encoding::Span>::{closure#0}
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_middle/src/util/bug.rs:35:34
  21:     0x7f4817d4a594 - rustc_middle[ebb8a7eed001d673]::ty::context::tls::with_opt::<rustc_middle[ebb8a7eed001d673]::util::bug::opt_span_bug_fmt<rustc_span[bb644a7833b5c27d]::span_encoding::Span>::{closure#0}, !>::{closure#0}
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_middle/src/ty/context/tls.rs:186:36
  22:     0x7f4817d4a548 - rustc_middle[ebb8a7eed001d673]::ty::context::tls::with_context_opt::<rustc_middle[ebb8a7eed001d673]::ty::context::tls::with_opt<rustc_middle[ebb8a7eed001d673]::util::bug::opt_span_bug_fmt<rustc_span[bb644a7833b5c27d]::span_encoding::Span>::{closure#0}, !>::{closure#0}, !>
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_middle/src/ty/context/tls.rs:132:18
  23:     0x7f4817d4bf77 - rustc_middle[ebb8a7eed001d673]::ty::context::tls::with_opt::<rustc_middle[ebb8a7eed001d673]::util::bug::opt_span_bug_fmt<rustc_span[bb644a7833b5c27d]::span_encoding::Span>::{closure#0}, !>
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_middle/src/ty/context/tls.rs:186:5
  24:     0x7f4817d4bf77 - rustc_middle[ebb8a7eed001d673]::util::bug::opt_span_bug_fmt::<rustc_span[bb644a7833b5c27d]::span_encoding::Span>
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_middle/src/util/bug.rs:31:5
  25:     0x7f4817d4bef3 - rustc_middle[ebb8a7eed001d673]::util::bug::bug_fmt
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_middle/src/util/bug.rs:15:5
  26:     0x7f4817cb1bdc - <rustc_middle[ebb8a7eed001d673]::ty::visit::ValidateBoundVars as rustc_middle[ebb8a7eed001d673]::ty::visit::TypeVisitor>::visit_region
  27:     0x7f4817c8661c - <rustc_middle[ebb8a7eed001d673]::ty::sty::Region as rustc_middle[ebb8a7eed001d673]::ty::visit::TypeVisitable>::visit_with::<rustc_middle[ebb8a7eed001d673]::ty::visit::ValidateBoundVars>
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_middle/src/ty/structural_impls.rs:737:9
  28:     0x7f4817c8661c - <rustc_middle[ebb8a7eed001d673]::ty::subst::GenericArg as rustc_middle[ebb8a7eed001d673]::ty::visit::TypeVisitable>::visit_with::<rustc_middle[ebb8a7eed001d673]::ty::visit::ValidateBoundVars>
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_middle/src/ty/subst.rs:243:45
  29:     0x7f4817c8661c - <&rustc_middle[ebb8a7eed001d673]::ty::list::List<rustc_middle[ebb8a7eed001d673]::ty::subst::GenericArg> as rustc_middle[ebb8a7eed001d673]::ty::visit::TypeVisitable>::visit_with::<rustc_middle[ebb8a7eed001d673]::ty::visit::ValidateBoundVars>::{closure#0}
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_middle/src/ty/subst.rs:541:38
  30:     0x7f4817c8661c - core[3a7799eb5b53fe42]::iter::traits::iterator::Iterator::try_for_each::call::<&rustc_middle[ebb8a7eed001d673]::ty::subst::GenericArg, core[3a7799eb5b53fe42]::ops::control_flow::ControlFlow<()>, <&rustc_middle[ebb8a7eed001d673]::ty::list::List<rustc_middle[ebb8a7eed001d673]::ty::subst::GenericArg> as rustc_middle[ebb8a7eed001d673]::ty::visit::TypeVisitable>::visit_with<rustc_middle[ebb8a7eed001d673]::ty::visit::ValidateBoundVars>::{closure#0}>::{closure#0}
                               at /home/matthias/vcs/github/rust_debug_assertions/library/core/src/iter/traits/iterator.rs:2320:26
  31:     0x7f4817c8661c - <core[3a7799eb5b53fe42]::slice::iter::Iter<rustc_middle[ebb8a7eed001d673]::ty::subst::GenericArg> as core[3a7799eb5b53fe42]::iter::traits::iterator::Iterator>::try_fold::<(), core[3a7799eb5b53fe42]::iter::traits::iterator::Iterator::try_for_each::call<&rustc_middle[ebb8a7eed001d673]::ty::subst::GenericArg, core[3a7799eb5b53fe42]::ops::control_flow::ControlFlow<()>, <&rustc_middle[ebb8a7eed001d673]::ty::list::List<rustc_middle[ebb8a7eed001d673]::ty::subst::GenericArg> as rustc_middle[ebb8a7eed001d673]::ty::visit::TypeVisitable>::visit_with<rustc_middle[ebb8a7eed001d673]::ty::visit::ValidateBoundVars>::{closure#0}>::{closure#0}, core[3a7799eb5b53fe42]::ops::control_flow::ControlFlow<()>>
                               at /home/matthias/vcs/github/rust_debug_assertions/library/core/src/iter/traits/iterator.rs:2262:21
  32:     0x7f4817c8661c - <core[3a7799eb5b53fe42]::slice::iter::Iter<rustc_middle[ebb8a7eed001d673]::ty::subst::GenericArg> as core[3a7799eb5b53fe42]::iter::traits::iterator::Iterator>::try_for_each::<<&rustc_middle[ebb8a7eed001d673]::ty::list::List<rustc_middle[ebb8a7eed001d673]::ty::subst::GenericArg> as rustc_middle[ebb8a7eed001d673]::ty::visit::TypeVisitable>::visit_with<rustc_middle[ebb8a7eed001d673]::ty::visit::ValidateBoundVars>::{closure#0}, core[3a7799eb5b53fe42]::ops::control_flow::ControlFlow<()>>
                               at /home/matthias/vcs/github/rust_debug_assertions/library/core/src/iter/traits/iterator.rs:2323:9
  33:     0x7f4817c8661c - <&rustc_middle[ebb8a7eed001d673]::ty::list::List<rustc_middle[ebb8a7eed001d673]::ty::subst::GenericArg> as rustc_middle[ebb8a7eed001d673]::ty::visit::TypeVisitable>::visit_with::<rustc_middle[ebb8a7eed001d673]::ty::visit::ValidateBoundVars>
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_middle/src/ty/subst.rs:541:21
  34:     0x7f4817c8661c - <rustc_middle[ebb8a7eed001d673]::ty::sty::AliasTy as rustc_middle[ebb8a7eed001d673]::ty::visit::TypeVisitable>::visit_with::<rustc_middle[ebb8a7eed001d673]::ty::visit::ValidateBoundVars>
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_middle/src/ty/sty.rs:1226:36
  35:     0x7f4817cb1827 - <rustc_middle[ebb8a7eed001d673]::ty::visit::ValidateBoundVars as rustc_middle[ebb8a7eed001d673]::ty::visit::TypeVisitor>::visit_ty
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_middle/src/ty/visit.rs:427:9
  36:     0x7f4817355a45 - <rustc_middle[ebb8a7eed001d673]::ty::Ty as rustc_middle[ebb8a7eed001d673]::ty::visit::TypeVisitable>::visit_with::<rustc_middle[ebb8a7eed001d673]::ty::visit::ValidateBoundVars>
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_middle/src/ty/structural_impls.rs:631:9
  37:     0x7f4817355a45 - <rustc_middle[ebb8a7eed001d673]::ty::sty::Binder<rustc_middle[ebb8a7eed001d673]::ty::Ty>>::bind_with_vars
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_middle/src/ty/sty.rs:998:13
  38:     0x7f48170eb5aa - <rustc_hir_typeck[e174858e69a5a042]::fn_ctxt::FnCtxt>::suggest_missing_return_type
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_hir_typeck/src/fn_ctxt/suggestions.rs:744:26
  39:     0x7f48170e31a9 - <rustc_hir_typeck[e174858e69a5a042]::fn_ctxt::FnCtxt>::suggest_mismatched_types_on_tail
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_hir_typeck/src/fn_ctxt/suggestions.rs:68:39
  40:     0x7f48173901be - <rustc_hir_typeck[e174858e69a5a042]::coercion::CoerceMany<&rustc_hir[a96c2743a87f38f5]::hir::Expr>>::report_return_mismatched_types
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_hir_typeck/src/coercion.rs:1686:17
  41:     0x7f481739ab9a - <rustc_hir_typeck[e174858e69a5a042]::coercion::CoerceMany<&rustc_hir[a96c2743a87f38f5]::hir::Expr>>::coerce_inner
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_hir_typeck/src/coercion.rs:1551:31
  42:     0x7f48170e0721 - <rustc_hir_typeck[e174858e69a5a042]::fn_ctxt::FnCtxt>::check_block_with_expected::{closure#0}
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_hir_typeck/src/fn_ctxt/checks.rs:1472:17
  43:     0x7f48170e0721 - <rustc_hir_typeck[e174858e69a5a042]::fn_ctxt::FnCtxt>::with_breakable_ctxt::<<rustc_hir_typeck[e174858e69a5a042]::fn_ctxt::FnCtxt>::check_block_with_expected::{closure#0}, ()>
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_hir_typeck/src/fn_ctxt/_impl.rs:1429:22
  44:     0x7f48170e0721 - <rustc_hir_typeck[e174858e69a5a042]::fn_ctxt::FnCtxt>::check_block_with_expected
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_hir_typeck/src/fn_ctxt/checks.rs:1452:26
  45:     0x7f48171331a2 - <rustc_hir_typeck[e174858e69a5a042]::fn_ctxt::FnCtxt>::check_expr_kind
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_hir_typeck/src/expr.rs:328:41
  46:     0x7f48170b9a2d - <rustc_hir_typeck[e174858e69a5a042]::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args::{closure#0}
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_hir_typeck/src/expr.rs:235:18
  47:     0x7f48170b9a2d - stacker[12027ce31fbb92f0]::maybe_grow::<rustc_middle[ebb8a7eed001d673]::ty::Ty, <rustc_hir_typeck[e174858e69a5a042]::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args::{closure#0}>
                               at /home/matthias/.cargo/registry/src/github.com-1ecc6299db9ec823/stacker-0.1.15/src/lib.rs:55:9
  48:     0x7f48170b9a2d - rustc_data_structures[3154e4d7ef9f97f8]::stack::ensure_sufficient_stack::<rustc_middle[ebb8a7eed001d673]::ty::Ty, <rustc_hir_typeck[e174858e69a5a042]::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args::{closure#0}>
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_data_structures/src/stack.rs:17:5
  49:     0x7f48170b9a2d - <rustc_hir_typeck[e174858e69a5a042]::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_hir_typeck/src/expr.rs:231:18
  50:     0x7f48171320f5 - <rustc_hir_typeck[e174858e69a5a042]::fn_ctxt::FnCtxt>::check_expr_with_expectation
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_hir_typeck/src/expr.rs:185:9
  51:     0x7f481713290a - <rustc_hir_typeck[e174858e69a5a042]::fn_ctxt::FnCtxt>::check_expr_kind
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_hir_typeck/src/expr.rs:342:39
  52:     0x7f48170b9a2d - <rustc_hir_typeck[e174858e69a5a042]::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args::{closure#0}
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_hir_typeck/src/expr.rs:235:18
  53:     0x7f48170b9a2d - stacker[12027ce31fbb92f0]::maybe_grow::<rustc_middle[ebb8a7eed001d673]::ty::Ty, <rustc_hir_typeck[e174858e69a5a042]::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args::{closure#0}>
                               at /home/matthias/.cargo/registry/src/github.com-1ecc6299db9ec823/stacker-0.1.15/src/lib.rs:55:9
  54:     0x7f48170b9a2d - rustc_data_structures[3154e4d7ef9f97f8]::stack::ensure_sufficient_stack::<rustc_middle[ebb8a7eed001d673]::ty::Ty, <rustc_hir_typeck[e174858e69a5a042]::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args::{closure#0}>
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_data_structures/src/stack.rs:17:5
  55:     0x7f48170b9a2d - <rustc_hir_typeck[e174858e69a5a042]::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_hir_typeck/src/expr.rs:231:18
  56:     0x7f48171320f5 - <rustc_hir_typeck[e174858e69a5a042]::fn_ctxt::FnCtxt>::check_expr_with_expectation
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_hir_typeck/src/expr.rs:185:9
  57:     0x7f48170e0384 - <rustc_hir_typeck[e174858e69a5a042]::fn_ctxt::FnCtxt>::check_block_with_expected::{closure#0}::{closure#0}
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_hir_typeck/src/fn_ctxt/checks.rs:1459:50
  58:     0x7f48170e0384 - <core[3a7799eb5b53fe42]::option::Option<&&rustc_hir[a96c2743a87f38f5]::hir::Expr>>::map::<rustc_middle[ebb8a7eed001d673]::ty::Ty, <rustc_hir_typeck[e174858e69a5a042]::fn_ctxt::FnCtxt>::check_block_with_expected::{closure#0}::{closure#0}>
                               at /home/matthias/vcs/github/rust_debug_assertions/library/core/src/option.rs:972:29
  59:     0x7f48170e0384 - <rustc_hir_typeck[e174858e69a5a042]::fn_ctxt::FnCtxt>::check_block_with_expected::{closure#0}
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_hir_typeck/src/fn_ctxt/checks.rs:1459:32
  60:     0x7f48170e0384 - <rustc_hir_typeck[e174858e69a5a042]::fn_ctxt::FnCtxt>::with_breakable_ctxt::<<rustc_hir_typeck[e174858e69a5a042]::fn_ctxt::FnCtxt>::check_block_with_expected::{closure#0}, ()>
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_hir_typeck/src/fn_ctxt/_impl.rs:1429:22
  61:     0x7f48170e0384 - <rustc_hir_typeck[e174858e69a5a042]::fn_ctxt::FnCtxt>::check_block_with_expected
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_hir_typeck/src/fn_ctxt/checks.rs:1452:26
  62:     0x7f48171331a2 - <rustc_hir_typeck[e174858e69a5a042]::fn_ctxt::FnCtxt>::check_expr_kind
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_hir_typeck/src/expr.rs:328:41
  63:     0x7f48170b9a2d - <rustc_hir_typeck[e174858e69a5a042]::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args::{closure#0}
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_hir_typeck/src/expr.rs:235:18
  64:     0x7f48170b9a2d - stacker[12027ce31fbb92f0]::maybe_grow::<rustc_middle[ebb8a7eed001d673]::ty::Ty, <rustc_hir_typeck[e174858e69a5a042]::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args::{closure#0}>
                               at /home/matthias/.cargo/registry/src/github.com-1ecc6299db9ec823/stacker-0.1.15/src/lib.rs:55:9
  65:     0x7f48170b9a2d - rustc_data_structures[3154e4d7ef9f97f8]::stack::ensure_sufficient_stack::<rustc_middle[ebb8a7eed001d673]::ty::Ty, <rustc_hir_typeck[e174858e69a5a042]::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args::{closure#0}>
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_data_structures/src/stack.rs:17:5
  66:     0x7f48170b9a2d - <rustc_hir_typeck[e174858e69a5a042]::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_hir_typeck/src/expr.rs:231:18
  67:     0x7f48171320f5 - <rustc_hir_typeck[e174858e69a5a042]::fn_ctxt::FnCtxt>::check_expr_with_expectation
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_hir_typeck/src/expr.rs:185:9
  68:     0x7f48170bb9a8 - <rustc_hir_typeck[e174858e69a5a042]::fn_ctxt::FnCtxt>::check_expr_with_hint
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_hir_typeck/src/expr.rs:137:9
  69:     0x7f48170bb9a8 - <rustc_hir_typeck[e174858e69a5a042]::fn_ctxt::FnCtxt>::check_return_expr
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_hir_typeck/src/expr.rs:835:30
  70:     0x7f48172ed068 - rustc_hir_typeck[e174858e69a5a042]::check::check_fn
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_hir_typeck/src/check.rs:124:9
  71:     0x7f481712a363 - <rustc_hir_typeck[e174858e69a5a042]::fn_ctxt::FnCtxt>::check_closure::{closure#0}
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_hir_typeck/src/closure.rs:85:31
  72:     0x7f481712a363 - <rustc_hir_typeck[e174858e69a5a042]::fn_ctxt::FnCtxt>::check_closure
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_hir_typeck/src/closure.rs:74:19
  73:     0x7f481712a363 - <rustc_hir_typeck[e174858e69a5a042]::fn_ctxt::FnCtxt>::check_expr_closure
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_hir_typeck/src/closure.rs:63:9
  74:     0x7f48171328b9 - <rustc_hir_typeck[e174858e69a5a042]::fn_ctxt::FnCtxt>::check_expr_kind
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_hir_typeck/src/expr.rs:327:43
  75:     0x7f48170b9a2d - <rustc_hir_typeck[e174858e69a5a042]::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args::{closure#0}
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_hir_typeck/src/expr.rs:235:18
  76:     0x7f48170b9a2d - stacker[12027ce31fbb92f0]::maybe_grow::<rustc_middle[ebb8a7eed001d673]::ty::Ty, <rustc_hir_typeck[e174858e69a5a042]::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args::{closure#0}>
                               at /home/matthias/.cargo/registry/src/github.com-1ecc6299db9ec823/stacker-0.1.15/src/lib.rs:55:9
  77:     0x7f48170b9a2d - rustc_data_structures[3154e4d7ef9f97f8]::stack::ensure_sufficient_stack::<rustc_middle[ebb8a7eed001d673]::ty::Ty, <rustc_hir_typeck[e174858e69a5a042]::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args::{closure#0}>
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_data_structures/src/stack.rs:17:5
  78:     0x7f48170b9a2d - <rustc_hir_typeck[e174858e69a5a042]::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_hir_typeck/src/expr.rs:231:18
  79:     0x7f48171320f5 - <rustc_hir_typeck[e174858e69a5a042]::fn_ctxt::FnCtxt>::check_expr_with_expectation
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_hir_typeck/src/expr.rs:185:9
  80:     0x7f48170d6a85 - <rustc_hir_typeck[e174858e69a5a042]::fn_ctxt::FnCtxt>::check_argument_types::{closure#0}
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_hir_typeck/src/fn_ctxt/checks.rs:255:30
  81:     0x7f48170d6a85 - <rustc_hir_typeck[e174858e69a5a042]::fn_ctxt::FnCtxt>::check_argument_types
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_hir_typeck/src/fn_ctxt/checks.rs:373:34
  82:     0x7f481709e381 - <rustc_hir_typeck[e174858e69a5a042]::fn_ctxt::FnCtxt>::confirm_builtin_call
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_hir_typeck/src/callee.rs:450:9
  83:     0x7f481709c8e9 - <rustc_hir_typeck[e174858e69a5a042]::fn_ctxt::FnCtxt>::check_call
  84:     0x7f4817132de1 - <rustc_hir_typeck[e174858e69a5a042]::fn_ctxt::FnCtxt>::check_expr_kind
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_hir_typeck/src/expr.rs:329:45
  85:     0x7f48170b9a2d - <rustc_hir_typeck[e174858e69a5a042]::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args::{closure#0}
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_hir_typeck/src/expr.rs:235:18
  86:     0x7f48170b9a2d - stacker[12027ce31fbb92f0]::maybe_grow::<rustc_middle[ebb8a7eed001d673]::ty::Ty, <rustc_hir_typeck[e174858e69a5a042]::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args::{closure#0}>
                               at /home/matthias/.cargo/registry/src/github.com-1ecc6299db9ec823/stacker-0.1.15/src/lib.rs:55:9
  87:     0x7f48170b9a2d - rustc_data_structures[3154e4d7ef9f97f8]::stack::ensure_sufficient_stack::<rustc_middle[ebb8a7eed001d673]::ty::Ty, <rustc_hir_typeck[e174858e69a5a042]::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args::{closure#0}>
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_data_structures/src/stack.rs:17:5
  88:     0x7f48170b9a2d - <rustc_hir_typeck[e174858e69a5a042]::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_hir_typeck/src/expr.rs:231:18
  89:     0x7f48171320f5 - <rustc_hir_typeck[e174858e69a5a042]::fn_ctxt::FnCtxt>::check_expr_with_expectation
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_hir_typeck/src/expr.rs:185:9
  90:     0x7f48170bb9a8 - <rustc_hir_typeck[e174858e69a5a042]::fn_ctxt::FnCtxt>::check_expr_with_hint
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_hir_typeck/src/expr.rs:137:9
  91:     0x7f48170bb9a8 - <rustc_hir_typeck[e174858e69a5a042]::fn_ctxt::FnCtxt>::check_return_expr
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_hir_typeck/src/expr.rs:835:30
  92:     0x7f48172ed068 - rustc_hir_typeck[e174858e69a5a042]::check::check_fn
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_hir_typeck/src/check.rs:124:9
  93:     0x7f48172e7d05 - rustc_hir_typeck[e174858e69a5a042]::typeck_with_fallback::<rustc_hir_typeck[e174858e69a5a042]::typeck::{closure#0}>::{closure#0}::{closure#1}
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_hir_typeck/src/lib.rs:219:13
  94:     0x7f48172e7d05 - <rustc_hir_typeck[e174858e69a5a042]::inherited::InheritedBuilder>::enter::<rustc_hir_typeck[e174858e69a5a042]::typeck_with_fallback<rustc_hir_typeck[e174858e69a5a042]::typeck::{closure#0}>::{closure#0}::{closure#1}, &rustc_middle[ebb8a7eed001d673]::ty::typeck_results::TypeckResults>
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_hir_typeck/src/inherited.rs:108:9
  95:     0x7f48172e7d05 - rustc_hir_typeck[e174858e69a5a042]::typeck_with_fallback::<rustc_hir_typeck[e174858e69a5a042]::typeck::{closure#0}>::{closure#0}
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_hir_typeck/src/lib.rs:202:26
  96:     0x7f48172e7d05 - rustc_hir_typeck[e174858e69a5a042]::typeck_with_fallback::<rustc_hir_typeck[e174858e69a5a042]::typeck::{closure#0}>
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_hir_typeck/src/lib.rs:185:36
  97:     0x7f48172e7d05 - rustc_hir_typeck[e174858e69a5a042]::typeck
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_hir_typeck/src/lib.rs:166:9
  98:     0x7f48187cd900 - rustc_query_system[a5ca8acfaca2cdd1]::query::plumbing::execute_job::<rustc_query_impl[c3bbf335af9135f0]::queries::typeck, rustc_query_impl[c3bbf335af9135f0]::plumbing::QueryCtxt>::{closure#0}
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_query_system/src/query/plumbing.rs:430:13
  99:     0x7f48187cd900 - stacker[12027ce31fbb92f0]::maybe_grow::<&rustc_middle[ebb8a7eed001d673]::ty::typeck_results::TypeckResults, rustc_query_system[a5ca8acfaca2cdd1]::query::plumbing::execute_job<rustc_query_impl[c3bbf335af9135f0]::queries::typeck, rustc_query_impl[c3bbf335af9135f0]::plumbing::QueryCtxt>::{closure#0}>
                               at /home/matthias/.cargo/registry/src/github.com-1ecc6299db9ec823/stacker-0.1.15/src/lib.rs:55:9
 100:     0x7f48187cd900 - rustc_data_structures[3154e4d7ef9f97f8]::stack::ensure_sufficient_stack::<&rustc_middle[ebb8a7eed001d673]::ty::typeck_results::TypeckResults, rustc_query_system[a5ca8acfaca2cdd1]::query::plumbing::execute_job<rustc_query_impl[c3bbf335af9135f0]::queries::typeck, rustc_query_impl[c3bbf335af9135f0]::plumbing::QueryCtxt>::{closure#0}>
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_data_structures/src/stack.rs:17:5
 101:     0x7f48187cd900 - <rustc_query_impl[c3bbf335af9135f0]::plumbing::QueryCtxt as rustc_query_system[a5ca8acfaca2cdd1]::query::QueryContext>::start_query::<&rustc_middle[ebb8a7eed001d673]::ty::typeck_results::TypeckResults, rustc_query_system[a5ca8acfaca2cdd1]::query::plumbing::execute_job<rustc_query_impl[c3bbf335af9135f0]::queries::typeck, rustc_query_impl[c3bbf335af9135f0]::plumbing::QueryCtxt>::{closure#0}>::{closure#0}::{closure#0}
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_query_impl/src/plumbing.rs:128:17
 102:     0x7f48187cd900 - rustc_middle[ebb8a7eed001d673]::ty::context::tls::enter_context::<<rustc_query_impl[c3bbf335af9135f0]::plumbing::QueryCtxt as rustc_query_system[a5ca8acfaca2cdd1]::query::QueryContext>::start_query<&rustc_middle[ebb8a7eed001d673]::ty::typeck_results::TypeckResults, rustc_query_system[a5ca8acfaca2cdd1]::query::plumbing::execute_job<rustc_query_impl[c3bbf335af9135f0]::queries::typeck, rustc_query_impl[c3bbf335af9135f0]::plumbing::QueryCtxt>::{closure#0}>::{closure#0}::{closure#0}, &rustc_middle[ebb8a7eed001d673]::ty::typeck_results::TypeckResults>::{closure#0}
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_middle/src/ty/context/tls.rs:115:38
 103:     0x7f48187cd900 - rustc_middle[ebb8a7eed001d673]::ty::context::tls::tlv::with_tlv::<rustc_middle[ebb8a7eed001d673]::ty::context::tls::enter_context<<rustc_query_impl[c3bbf335af9135f0]::plumbing::QueryCtxt as rustc_query_system[a5ca8acfaca2cdd1]::query::QueryContext>::start_query<&rustc_middle[ebb8a7eed001d673]::ty::typeck_results::TypeckResults, rustc_query_system[a5ca8acfaca2cdd1]::query::plumbing::execute_job<rustc_query_impl[c3bbf335af9135f0]::queries::typeck, rustc_query_impl[c3bbf335af9135f0]::plumbing::QueryCtxt>::{closure#0}>::{closure#0}::{closure#0}, &rustc_middle[ebb8a7eed001d673]::ty::typeck_results::TypeckResults>::{closure#0}, &rustc_middle[ebb8a7eed001d673]::ty::typeck_results::TypeckResults>
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_middle/src/ty/context/tls.rs:95:9
 104:     0x7f48187cd900 - rustc_middle[ebb8a7eed001d673]::ty::context::tls::enter_context::<<rustc_query_impl[c3bbf335af9135f0]::plumbing::QueryCtxt as rustc_query_system[a5ca8acfaca2cdd1]::query::QueryContext>::start_query<&rustc_middle[ebb8a7eed001d673]::ty::typeck_results::TypeckResults, rustc_query_system[a5ca8acfaca2cdd1]::query::plumbing::execute_job<rustc_query_impl[c3bbf335af9135f0]::queries::typeck, rustc_query_impl[c3bbf335af9135f0]::plumbing::QueryCtxt>::{closure#0}>::{closure#0}::{closure#0}, &rustc_middle[ebb8a7eed001d673]::ty::typeck_results::TypeckResults>
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_middle/src/ty/context/tls.rs:115:5
 105:     0x7f48187cd900 - <rustc_query_impl[c3bbf335af9135f0]::plumbing::QueryCtxt as rustc_query_system[a5ca8acfaca2cdd1]::query::QueryContext>::start_query::<&rustc_middle[ebb8a7eed001d673]::ty::typeck_results::TypeckResults, rustc_query_system[a5ca8acfaca2cdd1]::query::plumbing::execute_job<rustc_query_impl[c3bbf335af9135f0]::queries::typeck, rustc_query_impl[c3bbf335af9135f0]::plumbing::QueryCtxt>::{closure#0}>::{closure#0}
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_query_impl/src/plumbing.rs:127:13
 106:     0x7f48187cd900 - rustc_middle[ebb8a7eed001d673]::ty::context::tls::with_related_context::<<rustc_query_impl[c3bbf335af9135f0]::plumbing::QueryCtxt as rustc_query_system[a5ca8acfaca2cdd1]::query::QueryContext>::start_query<&rustc_middle[ebb8a7eed001d673]::ty::typeck_results::TypeckResults, rustc_query_system[a5ca8acfaca2cdd1]::query::plumbing::execute_job<rustc_query_impl[c3bbf335af9135f0]::queries::typeck, rustc_query_impl[c3bbf335af9135f0]::plumbing::QueryCtxt>::{closure#0}>::{closure#0}, &rustc_middle[ebb8a7eed001d673]::ty::typeck_results::TypeckResults>::{closure#0}
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_middle/src/ty/context/tls.rs:165:9
 107:     0x7f48187cd900 - rustc_middle[ebb8a7eed001d673]::ty::context::tls::with_context::<rustc_middle[ebb8a7eed001d673]::ty::context::tls::with_related_context<<rustc_query_impl[c3bbf335af9135f0]::plumbing::QueryCtxt as rustc_query_system[a5ca8acfaca2cdd1]::query::QueryContext>::start_query<&rustc_middle[ebb8a7eed001d673]::ty::typeck_results::TypeckResults, rustc_query_system[a5ca8acfaca2cdd1]::query::plumbing::execute_job<rustc_query_impl[c3bbf335af9135f0]::queries::typeck, rustc_query_impl[c3bbf335af9135f0]::plumbing::QueryCtxt>::{closure#0}>::{closure#0}, &rustc_middle[ebb8a7eed001d673]::ty::typeck_results::TypeckResults>::{closure#0}, &rustc_middle[ebb8a7eed001d673]::ty::typeck_results::TypeckResults>::{closure#0}
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_middle/src/ty/context/tls.rs:143:36
 108:     0x7f48187cd900 - rustc_middle[ebb8a7eed001d673]::ty::context::tls::with_context_opt::<rustc_middle[ebb8a7eed001d673]::ty::context::tls::with_context<rustc_middle[ebb8a7eed001d673]::ty::context::tls::with_related_context<<rustc_query_impl[c3bbf335af9135f0]::plumbing::QueryCtxt as rustc_query_system[a5ca8acfaca2cdd1]::query::QueryContext>::start_query<&rustc_middle[ebb8a7eed001d673]::ty::typeck_results::TypeckResults, rustc_query_system[a5ca8acfaca2cdd1]::query::plumbing::execute_job<rustc_query_impl[c3bbf335af9135f0]::queries::typeck, rustc_query_impl[c3bbf335af9135f0]::plumbing::QueryCtxt>::{closure#0}>::{closure#0}, &rustc_middle[ebb8a7eed001d673]::ty::typeck_results::TypeckResults>::{closure#0}, &rustc_middle[ebb8a7eed001d673]::ty::typeck_results::TypeckResults>::{closure#0}, &rustc_middle[ebb8a7eed001d673]::ty::typeck_results::TypeckResults>
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_middle/src/ty/context/tls.rs:132:18
 109:     0x7f48187cd900 - rustc_middle[ebb8a7eed001d673]::ty::context::tls::with_context::<rustc_middle[ebb8a7eed001d673]::ty::context::tls::with_related_context<<rustc_query_impl[c3bbf335af9135f0]::plumbing::QueryCtxt as rustc_query_system[a5ca8acfaca2cdd1]::query::QueryContext>::start_query<&rustc_middle[ebb8a7eed001d673]::ty::typeck_results::TypeckResults, rustc_query_system[a5ca8acfaca2cdd1]::query::plumbing::execute_job<rustc_query_impl[c3bbf335af9135f0]::queries::typeck, rustc_query_impl[c3bbf335af9135f0]::plumbing::QueryCtxt>::{closure#0}>::{closure#0}, &rustc_middle[ebb8a7eed001d673]::ty::typeck_results::TypeckResults>::{closure#0}, &rustc_middle[ebb8a7eed001d673]::ty::typeck_results::TypeckResults>
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_middle/src/ty/context/tls.rs:143:5
 110:     0x7f48187cd900 - rustc_middle[ebb8a7eed001d673]::ty::context::tls::with_related_context::<<rustc_query_impl[c3bbf335af9135f0]::plumbing::QueryCtxt as rustc_query_system[a5ca8acfaca2cdd1]::query::QueryContext>::start_query<&rustc_middle[ebb8a7eed001d673]::ty::typeck_results::TypeckResults, rustc_query_system[a5ca8acfaca2cdd1]::query::plumbing::execute_job<rustc_query_impl[c3bbf335af9135f0]::queries::typeck, rustc_query_impl[c3bbf335af9135f0]::plumbing::QueryCtxt>::{closure#0}>::{closure#0}, &rustc_middle[ebb8a7eed001d673]::ty::typeck_results::TypeckResults>
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_middle/src/ty/context/tls.rs:156:5
 111:     0x7f48187cd900 - <rustc_query_impl[c3bbf335af9135f0]::plumbing::QueryCtxt as rustc_query_system[a5ca8acfaca2cdd1]::query::QueryContext>::start_query::<&rustc_middle[ebb8a7eed001d673]::ty::typeck_results::TypeckResults, rustc_query_system[a5ca8acfaca2cdd1]::query::plumbing::execute_job<rustc_query_impl[c3bbf335af9135f0]::queries::typeck, rustc_query_impl[c3bbf335af9135f0]::plumbing::QueryCtxt>::{closure#0}>
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_query_impl/src/plumbing.rs:112:9
 112:     0x7f48187cd900 - rustc_query_system[a5ca8acfaca2cdd1]::query::plumbing::execute_job::<rustc_query_impl[c3bbf335af9135f0]::queries::typeck, rustc_query_impl[c3bbf335af9135f0]::plumbing::QueryCtxt>
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_query_system/src/query/plumbing.rs:429:22
 113:     0x7f48187cd900 - rustc_query_system[a5ca8acfaca2cdd1]::query::plumbing::try_execute_query::<rustc_query_impl[c3bbf335af9135f0]::queries::typeck, rustc_query_impl[c3bbf335af9135f0]::plumbing::QueryCtxt>
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_query_system/src/query/plumbing.rs:369:17
 114:     0x7f4818a2bae3 - rustc_query_system[a5ca8acfaca2cdd1]::query::plumbing::get_query::<rustc_query_impl[c3bbf335af9135f0]::queries::typeck, rustc_query_impl[c3bbf335af9135f0]::plumbing::QueryCtxt, rustc_middle[ebb8a7eed001d673]::dep_graph::dep_node::DepKind>
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_query_system/src/query/plumbing.rs:747:36
 115:     0x7f4818a2bae3 - <rustc_query_impl[c3bbf335af9135f0]::Queries as rustc_middle[ebb8a7eed001d673]::ty::query::QueryEngine>::typeck::{closure#0}
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_query_impl/src/plumbing.rs:730:17
 116:     0x7f4818a2bae3 - <rustc_query_impl[c3bbf335af9135f0]::Queries as rustc_middle[ebb8a7eed001d673]::ty::query::QueryEngine>::typeck
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_query_impl/src/lib.rs:44:1
 117:     0x7f4817b44ddb - <rustc_middle[ebb8a7eed001d673]::ty::query::TyCtxtAt>::typeck::<rustc_span[bb644a7833b5c27d]::def_id::LocalDefId>
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_middle/src/ty/query.rs:252:29
 118:     0x7f4817b44ddb - <rustc_middle[ebb8a7eed001d673]::ty::context::TyCtxt>::typeck::<rustc_span[bb644a7833b5c27d]::def_id::LocalDefId>
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_middle/src/ty/query.rs:238:35
 119:     0x7f4817b44ddb - <rustc_middle[ebb8a7eed001d673]::ty::context::TyCtxt>::typeck_opt_const_arg
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_middle/src/ty/context.rs:501:18
 120:     0x7f4817ea648f - <rustc_mir_build[f8fee0aca10cdf74]::thir::cx::Cx>::new
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_mir_build/src/thir/cx/mod.rs:99:30
 121:     0x7f4817ea648f - rustc_mir_build[f8fee0aca10cdf74]::thir::cx::thir_body
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_mir_build/src/thir/cx/mod.rs:27:18
 122:     0x7f48187e1954 - rustc_query_system[a5ca8acfaca2cdd1]::query::plumbing::execute_job::<rustc_query_impl[c3bbf335af9135f0]::queries::thir_body, rustc_query_impl[c3bbf335af9135f0]::plumbing::QueryCtxt>::{closure#0}
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_query_system/src/query/plumbing.rs:430:13
 123:     0x7f48187e1954 - stacker[12027ce31fbb92f0]::maybe_grow::<core[3a7799eb5b53fe42]::result::Result<(&rustc_data_structures[3154e4d7ef9f97f8]::steal::Steal<rustc_middle[ebb8a7eed001d673]::thir::Thir>, rustc_middle[ebb8a7eed001d673]::thir::ExprId), rustc_errors[7deb60bbfdd1de15]::ErrorGuaranteed>, rustc_query_system[a5ca8acfaca2cdd1]::query::plumbing::execute_job<rustc_query_impl[c3bbf335af9135f0]::queries::thir_body, rustc_query_impl[c3bbf335af9135f0]::plumbing::QueryCtxt>::{closure#0}>
                               at /home/matthias/.cargo/registry/src/github.com-1ecc6299db9ec823/stacker-0.1.15/src/lib.rs:55:9
 124:     0x7f48187e1954 - rustc_data_structures[3154e4d7ef9f97f8]::stack::ensure_sufficient_stack::<core[3a7799eb5b53fe42]::result::Result<(&rustc_data_structures[3154e4d7ef9f97f8]::steal::Steal<rustc_middle[ebb8a7eed001d673]::thir::Thir>, rustc_middle[ebb8a7eed001d673]::thir::ExprId), rustc_errors[7deb60bbfdd1de15]::ErrorGuaranteed>, rustc_query_system[a5ca8acfaca2cdd1]::query::plumbing::execute_job<rustc_query_impl[c3bbf335af9135f0]::queries::thir_body, rustc_query_impl[c3bbf335af9135f0]::plumbing::QueryCtxt>::{closure#0}>
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_data_structures/src/stack.rs:17:5
 125:     0x7f48187e1954 - <rustc_query_impl[c3bbf335af9135f0]::plumbing::QueryCtxt as rustc_query_system[a5ca8acfaca2cdd1]::query::QueryContext>::start_query::<core[3a7799eb5b53fe42]::result::Result<(&rustc_data_structures[3154e4d7ef9f97f8]::steal::Steal<rustc_middle[ebb8a7eed001d673]::thir::Thir>, rustc_middle[ebb8a7eed001d673]::thir::ExprId), rustc_errors[7deb60bbfdd1de15]::ErrorGuaranteed>, rustc_query_system[a5ca8acfaca2cdd1]::query::plumbing::execute_job<rustc_query_impl[c3bbf335af9135f0]::queries::thir_body, rustc_query_impl[c3bbf335af9135f0]::plumbing::QueryCtxt>::{closure#0}>::{closure#0}::{closure#0}
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_query_impl/src/plumbing.rs:128:17
 126:     0x7f48187e1954 - rustc_middle[ebb8a7eed001d673]::ty::context::tls::enter_context::<<rustc_query_impl[c3bbf335af9135f0]::plumbing::QueryCtxt as rustc_query_system[a5ca8acfaca2cdd1]::query::QueryContext>::start_query<core[3a7799eb5b53fe42]::result::Result<(&rustc_data_structures[3154e4d7ef9f97f8]::steal::Steal<rustc_middle[ebb8a7eed001d673]::thir::Thir>, rustc_middle[ebb8a7eed001d673]::thir::ExprId), rustc_errors[7deb60bbfdd1de15]::ErrorGuaranteed>, rustc_query_system[a5ca8acfaca2cdd1]::query::plumbing::execute_job<rustc_query_impl[c3bbf335af9135f0]::queries::thir_body, rustc_query_impl[c3bbf335af9135f0]::plumbing::QueryCtxt>::{closure#0}>::{closure#0}::{closure#0}, core[3a7799eb5b53fe42]::result::Result<(&rustc_data_structures[3154e4d7ef9f97f8]::steal::Steal<rustc_middle[ebb8a7eed001d673]::thir::Thir>, rustc_middle[ebb8a7eed001d673]::thir::ExprId), rustc_errors[7deb60bbfdd1de15]::ErrorGuaranteed>>::{closure#0}
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_middle/src/ty/context/tls.rs:115:38
 127:     0x7f48187e1954 - rustc_middle[ebb8a7eed001d673]::ty::context::tls::tlv::with_tlv::<rustc_middle[ebb8a7eed001d673]::ty::context::tls::enter_context<<rustc_query_impl[c3bbf335af9135f0]::plumbing::QueryCtxt as rustc_query_system[a5ca8acfaca2cdd1]::query::QueryContext>::start_query<core[3a7799eb5b53fe42]::result::Result<(&rustc_data_structures[3154e4d7ef9f97f8]::steal::Steal<rustc_middle[ebb8a7eed001d673]::thir::Thir>, rustc_middle[ebb8a7eed001d673]::thir::ExprId), rustc_errors[7deb60bbfdd1de15]::ErrorGuaranteed>, rustc_query_system[a5ca8acfaca2cdd1]::query::plumbing::execute_job<rustc_query_impl[c3bbf335af9135f0]::queries::thir_body, rustc_query_impl[c3bbf335af9135f0]::plumbing::QueryCtxt>::{closure#0}>::{closure#0}::{closure#0}, core[3a7799eb5b53fe42]::result::Result<(&rustc_data_structures[3154e4d7ef9f97f8]::steal::Steal<rustc_middle[ebb8a7eed001d673]::thir::Thir>, rustc_middle[ebb8a7eed001d673]::thir::ExprId), rustc_errors[7deb60bbfdd1de15]::ErrorGuaranteed>>::{closure#0}, core[3a7799eb5b53fe42]::result::Result<(&rustc_data_structures[3154e4d7ef9f97f8]::steal::Steal<rustc_middle[ebb8a7eed001d673]::thir::Thir>, rustc_middle[ebb8a7eed001d673]::thir::ExprId), rustc_errors[7deb60bbfdd1de15]::ErrorGuaranteed>>
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_middle/src/ty/context/tls.rs:95:9
 128:     0x7f48187e1954 - rustc_middle[ebb8a7eed001d673]::ty::context::tls::enter_context::<<rustc_query_impl[c3bbf335af9135f0]::plumbing::QueryCtxt as rustc_query_system[a5ca8acfaca2cdd1]::query::QueryContext>::start_query<core[3a7799eb5b53fe42]::result::Result<(&rustc_data_structures[3154e4d7ef9f97f8]::steal::Steal<rustc_middle[ebb8a7eed001d673]::thir::Thir>, rustc_middle[ebb8a7eed001d673]::thir::ExprId), rustc_errors[7deb60bbfdd1de15]::ErrorGuaranteed>, rustc_query_system[a5ca8acfaca2cdd1]::query::plumbing::execute_job<rustc_query_impl[c3bbf335af9135f0]::queries::thir_body, rustc_query_impl[c3bbf335af9135f0]::plumbing::QueryCtxt>::{closure#0}>::{closure#0}::{closure#0}, core[3a7799eb5b53fe42]::result::Result<(&rustc_data_structures[3154e4d7ef9f97f8]::steal::Steal<rustc_middle[ebb8a7eed001d673]::thir::Thir>, rustc_middle[ebb8a7eed001d673]::thir::ExprId), rustc_errors[7deb60bbfdd1de15]::ErrorGuaranteed>>
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_middle/src/ty/context/tls.rs:115:5
 129:     0x7f48187e1954 - <rustc_query_impl[c3bbf335af9135f0]::plumbing::QueryCtxt as rustc_query_system[a5ca8acfaca2cdd1]::query::QueryContext>::start_query::<core[3a7799eb5b53fe42]::result::Result<(&rustc_data_structures[3154e4d7ef9f97f8]::steal::Steal<rustc_middle[ebb8a7eed001d673]::thir::Thir>, rustc_middle[ebb8a7eed001d673]::thir::ExprId), rustc_errors[7deb60bbfdd1de15]::ErrorGuaranteed>, rustc_query_system[a5ca8acfaca2cdd1]::query::plumbing::execute_job<rustc_query_impl[c3bbf335af9135f0]::queries::thir_body, rustc_query_impl[c3bbf335af9135f0]::plumbing::QueryCtxt>::{closure#0}>::{closure#0}
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_query_impl/src/plumbing.rs:127:13
 130:     0x7f48187e1954 - rustc_middle[ebb8a7eed001d673]::ty::context::tls::with_related_context::<<rustc_query_impl[c3bbf335af9135f0]::plumbing::QueryCtxt as rustc_query_system[a5ca8acfaca2cdd1]::query::QueryContext>::start_query<core[3a7799eb5b53fe42]::result::Result<(&rustc_data_structures[3154e4d7ef9f97f8]::steal::Steal<rustc_middle[ebb8a7eed001d673]::thir::Thir>, rustc_middle[ebb8a7eed001d673]::thir::ExprId), rustc_errors[7deb60bbfdd1de15]::ErrorGuaranteed>, rustc_query_system[a5ca8acfaca2cdd1]::query::plumbing::execute_job<rustc_query_impl[c3bbf335af9135f0]::queries::thir_body, rustc_query_impl[c3bbf335af9135f0]::plumbing::QueryCtxt>::{closure#0}>::{closure#0}, core[3a7799eb5b53fe42]::result::Result<(&rustc_data_structures[3154e4d7ef9f97f8]::steal::Steal<rustc_middle[ebb8a7eed001d673]::thir::Thir>, rustc_middle[ebb8a7eed001d673]::thir::ExprId), rustc_errors[7deb60bbfdd1de15]::ErrorGuaranteed>>::{closure#0}
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_middle/src/ty/context/tls.rs:165:9
 131:     0x7f48187e1954 - rustc_middle[ebb8a7eed001d673]::ty::context::tls::with_context::<rustc_middle[ebb8a7eed001d673]::ty::context::tls::with_related_context<<rustc_query_impl[c3bbf335af9135f0]::plumbing::QueryCtxt as rustc_query_system[a5ca8acfaca2cdd1]::query::QueryContext>::start_query<core[3a7799eb5b53fe42]::result::Result<(&rustc_data_structures[3154e4d7ef9f97f8]::steal::Steal<rustc_middle[ebb8a7eed001d673]::thir::Thir>, rustc_middle[ebb8a7eed001d673]::thir::ExprId), rustc_errors[7deb60bbfdd1de15]::ErrorGuaranteed>, rustc_query_system[a5ca8acfaca2cdd1]::query::plumbing::execute_job<rustc_query_impl[c3bbf335af9135f0]::queries::thir_body, rustc_query_impl[c3bbf335af9135f0]::plumbing::QueryCtxt>::{closure#0}>::{closure#0}, core[3a7799eb5b53fe42]::result::Result<(&rustc_data_structures[3154e4d7ef9f97f8]::steal::Steal<rustc_middle[ebb8a7eed001d673]::thir::Thir>, rustc_middle[ebb8a7eed001d673]::thir::ExprId), rustc_errors[7deb60bbfdd1de15]::ErrorGuaranteed>>::{closure#0}, core[3a7799eb5b53fe42]::result::Result<(&rustc_data_structures[3154e4d7ef9f97f8]::steal::Steal<rustc_middle[ebb8a7eed001d673]::thir::Thir>, rustc_middle[ebb8a7eed001d673]::thir::ExprId), rustc_errors[7deb60bbfdd1de15]::ErrorGuaranteed>>::{closure#0}
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_middle/src/ty/context/tls.rs:143:36
 132:     0x7f48187e1954 - rustc_middle[ebb8a7eed001d673]::ty::context::tls::with_context_opt::<rustc_middle[ebb8a7eed001d673]::ty::context::tls::with_context<rustc_middle[ebb8a7eed001d673]::ty::context::tls::with_related_context<<rustc_query_impl[c3bbf335af9135f0]::plumbing::QueryCtxt as rustc_query_system[a5ca8acfaca2cdd1]::query::QueryContext>::start_query<core[3a7799eb5b53fe42]::result::Result<(&rustc_data_structures[3154e4d7ef9f97f8]::steal::Steal<rustc_middle[ebb8a7eed001d673]::thir::Thir>, rustc_middle[ebb8a7eed001d673]::thir::ExprId), rustc_errors[7deb60bbfdd1de15]::ErrorGuaranteed>, rustc_query_system[a5ca8acfaca2cdd1]::query::plumbing::execute_job<rustc_query_impl[c3bbf335af9135f0]::queries::thir_body, rustc_query_impl[c3bbf335af9135f0]::plumbing::QueryCtxt>::{closure#0}>::{closure#0}, core[3a7799eb5b53fe42]::result::Result<(&rustc_data_structures[3154e4d7ef9f97f8]::steal::Steal<rustc_middle[ebb8a7eed001d673]::thir::Thir>, rustc_middle[ebb8a7eed001d673]::thir::ExprId), rustc_errors[7deb60bbfdd1de15]::ErrorGuaranteed>>::{closure#0}, core[3a7799eb5b53fe42]::result::Result<(&rustc_data_structures[3154e4d7ef9f97f8]::steal::Steal<rustc_middle[ebb8a7eed001d673]::thir::Thir>, rustc_middle[ebb8a7eed001d673]::thir::ExprId), rustc_errors[7deb60bbfdd1de15]::ErrorGuaranteed>>::{closure#0}, core[3a7799eb5b53fe42]::result::Result<(&rustc_data_structures[3154e4d7ef9f97f8]::steal::Steal<rustc_middle[ebb8a7eed001d673]::thir::Thir>, rustc_middle[ebb8a7eed001d673]::thir::ExprId), rustc_errors[7deb60bbfdd1de15]::ErrorGuaranteed>>
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_middle/src/ty/context/tls.rs:132:18
 133:     0x7f48187e1954 - rustc_middle[ebb8a7eed001d673]::ty::context::tls::with_context::<rustc_middle[ebb8a7eed001d673]::ty::context::tls::with_related_context<<rustc_query_impl[c3bbf335af9135f0]::plumbing::QueryCtxt as rustc_query_system[a5ca8acfaca2cdd1]::query::QueryContext>::start_query<core[3a7799eb5b53fe42]::result::Result<(&rustc_data_structures[3154e4d7ef9f97f8]::steal::Steal<rustc_middle[ebb8a7eed001d673]::thir::Thir>, rustc_middle[ebb8a7eed001d673]::thir::ExprId), rustc_errors[7deb60bbfdd1de15]::ErrorGuaranteed>, rustc_query_system[a5ca8acfaca2cdd1]::query::plumbing::execute_job<rustc_query_impl[c3bbf335af9135f0]::queries::thir_body, rustc_query_impl[c3bbf335af9135f0]::plumbing::QueryCtxt>::{closure#0}>::{closure#0}, core[3a7799eb5b53fe42]::result::Result<(&rustc_data_structures[3154e4d7ef9f97f8]::steal::Steal<rustc_middle[ebb8a7eed001d673]::thir::Thir>, rustc_middle[ebb8a7eed001d673]::thir::ExprId), rustc_errors[7deb60bbfdd1de15]::ErrorGuaranteed>>::{closure#0}, core[3a7799eb5b53fe42]::result::Result<(&rustc_data_structures[3154e4d7ef9f97f8]::steal::Steal<rustc_middle[ebb8a7eed001d673]::thir::Thir>, rustc_middle[ebb8a7eed001d673]::thir::ExprId), rustc_errors[7deb60bbfdd1de15]::ErrorGuaranteed>>
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_middle/src/ty/context/tls.rs:143:5
 
 
<snip>


note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.69.0-dev running on x86_64-unknown-linux-gnu

query stack during panic:
#0 [typeck] type-checking `str`
#1 [thir_body] building THIR for `str`
#2 [mir_built] building MIR for `str`
#3 [unsafety_check_result] unsafety-checking `str`
#4 [mir_const] preparing `str` for borrow checking
#5 [mir_promoted] processing MIR for `str`
#6 [mir_borrowck] borrow-checking `str`
#7 [type_of] computing type of `str::{opaque#0}`
#8 [check_mod_item_types] checking item types in top-level module
#9 [analysis] running analysis passes on this crate
end of query stack
error: aborting due to 3 previous errors

Some errors have detailed explanations: E0601, E0670.
For more information about an error, try `rustc --explain E0601`.

```

</p>
</details>

--- a/compiler/rustc_hir_typeck/src/fn_ctxt/suggestions.rs
+++ b/compiler/rustc_hir_typeck/src/fn_ctxt/suggestions.rs
@@ -669,6 +669,7 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
     /// This routine checks if the return type is left as default, the method is not part of an
     /// `impl` block and that it isn't the `main` method. If so, it suggests setting the return
     /// type.
+    #[instrument(level = "trace", skip(self, err))]
     pub(in super::super) fn suggest_missing_return_type(
         &self,
         err: &mut Diagnostic,
@@ -705,28 +706,24 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                     return true
                 }
             }
-            hir::FnRetTy::Return(ty) => {
-                let span = ty.span;
+            hir::FnRetTy::Return(hir_ty) => {
+                let span = hir_ty.span;
 
-                if let hir::TyKind::OpaqueDef(item_id, ..) = ty.kind
-                && let hir::Node::Item(hir::Item {
-                    kind: hir::ItemKind::OpaqueTy(op_ty),
-                    ..
-                }) = self.tcx.hir().get(item_id.hir_id())
-                && let hir::OpaqueTy {
-                    bounds: [bound], ..
-                } = op_ty
-                && let hir::GenericBound::LangItemTrait(
-                    hir::LangItem::Future, _, _, generic_args) = bound
-                && let hir::GenericArgs { bindings: [ty_binding], .. } = generic_args
-                && let hir::TypeBinding { kind, .. } = ty_binding
-                && let hir::TypeBindingKind::Equality { term } = kind
-                && let hir::Term::Ty(term_ty) = term {
+                if let hir::TyKind::OpaqueDef(item_id, ..) = hir_ty.kind
+                    && let hir::Node::Item(hir::Item {
+                        kind: hir::ItemKind::OpaqueTy(op_ty),
+                        ..
+                    }) = self.tcx.hir().get(item_id.hir_id())
+                    && let [hir::GenericBound::LangItemTrait(
+                        hir::LangItem::Future, _, _, generic_args)] = op_ty.bounds
+                    && let hir::GenericArgs { bindings: [ty_binding], .. } = generic_args
+                    && let hir::TypeBindingKind::Equality { term: hir::Term::Ty(term) } = ty_binding.kind
+                {
                     // Check if async function's return type was omitted.
                     // Don't emit suggestions if the found type is `impl Future<...>`.
-                    debug!("suggest_missing_return_type: found = {:?}", found);
+                    debug!(?found);
                     if found.is_suggestable(self.tcx, false) {
-                        if term_ty.span.is_empty() {
+                        if term.span.is_empty() {
                             err.subdiagnostic(AddReturnTypeSuggestion::Add { span, found: found.to_string() });
                             return true;
                         } else {
@@ -737,11 +734,11 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
 
                 // Only point to return type if the expected type is the return type, as if they
                 // are not, the expectation must have been caused by something else.
-                debug!("suggest_missing_return_type: return type {:?} node {:?}", ty, ty.kind);
-                let ty = self.astconv().ast_ty_to_ty(ty);
-                debug!("suggest_missing_return_type: return type {:?}", ty);
-                debug!("suggest_missing_return_type: expected type {:?}", ty);
-                let bound_vars = self.tcx.late_bound_vars(fn_id);
+                debug!("return type {:?}", hir_ty);
+                let ty = self.astconv().ast_ty_to_ty(hir_ty);
+                debug!("return type {:?}", ty);
+                debug!("expected type {:?}", expected);
+                let bound_vars = self.tcx.late_bound_vars(hir_ty.hir_id.owner.into());
                 let ty = Binder::bind_with_vars(ty, bound_vars);
                 let ty = self.normalize(span, ty);
                 let ty = self.tcx.erase_late_bound_regions(ty);

--- a/tests/ui/suggestions/issue-107860.rs
+++ b/tests/ui/suggestions/issue-107860.rs
@@ -1,0 +1,6 @@
+// edition: 2021
+
+async fn str<T>(T: &str) -> &str { &str }
+//~^ ERROR mismatched types
+
+fn main() {}

--- a/tests/ui/suggestions/issue-107860.stderr
+++ b/tests/ui/suggestions/issue-107860.stderr
@@ -1,0 +1,12 @@
+error[E0308]: mismatched types
+  --> $DIR/issue-107860.rs:3:36
+   |
+LL | async fn str<T>(T: &str) -> &str { &str }
+   |                                    ^^^^ expected `&str`, found `&fn(&str) -> ... {str::<...>}`
+   |
+   = note: expected reference `&str`
+              found reference `&for<'a> fn(&'a str) -> impl Future<Output = &'a str> {str::<_>}`
+
+error: aborting due to previous error
+
+For more information about this error, try `rustc --explain E0308`.

